simulator/dataAnalysis/dataAnalysis.py

```diff
@@ -0,0 +1,99 @@
+import json
+import os
+
+# returns a list of json objects
+def create_json_list(json_file_name):
+    json_file = open(json_file_name)
+    list_form = json_file.read()[1:-1].split('},')
+    json_file.close()
+
+    list_of_stringified_json = [i + '}' for i in list_form[:-1]] + [list_form[-1]]
+    return [json.loads(x) for x in list_of_stringified_json]
+
+
+# get number of nodes added to the network
+def get_num_nodes(json_list):
+    num_nodes = 0
+    for event in json_list:
+        if event['kind'] == 'add-node':
+            num_nodes += 1
+    return num_nodes
+
+
+# get min, max, median block propagation times
+def get_propagation_info(json_list):
+    num_nodes = get_num_nodes(json_list)
+
+    # stores block_id: [prop_start_time, prop_end_time]
+    block_propagation_times = {}
+    # stores block_id: {nodes who've seen the block}
+    block_propagation = {}
+    for event in json_list:
+        if event['kind'] == 'add-block':
+            block_id = event['content']['block-id']
+            node_id = event['content']['node-id']
+
+            # if this block has been seen by a node before
+            if block_id in block_propagation.keys():
+                # add the new node in the list of receivers
+                block_propagation[block_id].add(node_id)
+                # the first time the block is seen by all nodes, mark the propagation end timestamp
+                if len(block_propagation[block_id]) == num_nodes and len(block_propagation_times[block_id]) < 2:
```

Comment:
is `len(block_propagation_times[block_id])==1` better? and why do you need to check that?

Reply:
Yes; switched to ==1. I believe a block can be sent to a node that has already seen it, so we want to log the first end time. I'll double check that that is the case.

Reply:
That assumption makes sense, so its good to have ==1